Allow customization of transaction initiation statement

JonasProgrammer opened the pull request:
- model postgres transaction options in code
- add options parameter to long-form of begin_op::operator()
- add .with_options()-helper to begin_op to control the implicit options

All credit for the interface ideas goes to @thed636

thed636 commented on include/ozo/transaction_options.h:
```diff
@@ -0,0 +1,76 @@
+    constexpr static struct read_committed_t {
+    } read_committed{};
+    constexpr static struct read_uncommitted_t {
+    } read_uncommitted{};
```
It is better to use `boost::hana::type` because in this case, you will get Comparable constants

```c++
struct isolation_level {
    constexpr static hana::type<class serializable_tag> serializable{};
    constexpr static hana::type<class repeatable_read_tag> repeatable_read{};
    constexpr static hana::type<class read_committed_tag> read_committed{};
    constexpr static hana::type<class read_uncommitted_tag> read_uncommitted{};
};
```

thed636 commented on include/ozo/detail/begin_statement_builder.h:
```diff
@@ -0,0 +1,85 @@
+namespace ozo::detail {
+
+struct begin_statement_builder {
+    constexpr static auto empty_sql() {
```
Maybe the constant is better?

```c++
static constexpr auto empty_sql = hana::string_c<>;
```

Since there is no parameter in your query, it is ok to use `hana::string` for the text part and then use `ozo::make_query(text)` to create an open transaction query.

thed636 commented on include/ozo/detail/begin_statement_builder.h:
```diff
@@ -0,0 +1,85 @@
+        using namespace ozo::literals;
+
+        auto prefix = " ISOLATION LEVEL "_SQL;
+        if constexpr (std::is_same_v<Level, isolation_level::serializable_t>) {
```
So if `isolation_level` entries are defined as `hana::type_c`, you can use `operator ==` like this:

```c++
template <typename Level>
constexpr static auto level_part(Level level) {
    const auto prefix = BOOST_HANA_STRING(" ISOLATION LEVEL ");
    if constexpr (decltype(level==none)::value) {
        return empty_sql;
    } else if constexpr (decltype(level == isolation_level::serializable)::value) {
        return prefix + BOOST_HANA_STRING("SERIALIZABLE");
   //...
    } else {
        // For the unknown type we will generate an error message
        static_assert(std::is_void_v<Level>, "unknown isolation level");
    }
}
```

thed636 commented on include/ozo/detail/begin_statement_builder.h:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    template <typename Mode, typename =  std::enable_if_t<is_transaction_mode_v<Mode>>>
+    constexpr static auto mode_part(Mode) {
```
The same thing here:

```c++
template <typename Mode>
constexpr static auto mode_part(Mode mode) {
    if constexpr (decltype(mode == none)::value) {
        return empty_sql;
    } else if constexpr (decltype(mode == transaction_mode::read_write)::value) {
       return BOOST_HANA_STRING(" READ WRITE");
    //...
    } else {
        static_assert(std::is_void_v<Mode>, "unknown transaction mode");
    }
}
```

thed636 commented on include/ozo/transaction.h:
```diff
@@ -54,17 +84,17 @@ struct commit_op {
-    template <typename T, typename CompletionToken>
-    auto operator() (impl::transaction<T>&& transaction, CompletionToken&& token) const {
+    template <typename T, typename Options, typename CompletionToken>
+    auto operator() (impl::transaction<T, Options>&& transaction, CompletionToken&& token) const {
```
I think it is ok to use variadic params for the transaction, like:

```c++
template <typename CompletionToken, typename ...Ts>
auto operator() (impl::transaction<Ts...>&& transaction, CompletionToken&& token) const
```

thed636 commented on include/ozo/transaction_options.h:
```diff
@@ -0,0 +1,76 @@
+};
+
+template <typename T>
+constexpr bool is_isolation_level_v = std::is_same_v<T, isolation_level::serializable_t> ||
```
And you do not need such things like this.

thed636 commented on include/ozo/detail/begin_statement_builder.h:
```diff
@@ -0,0 +1,85 @@
+            return " NOT"_SQL + std::move(deferrable_sql);
+        } else {
+            return deferrable_sql;
+        }
```
In the same way there maybe something like this:

```c++
template <typename Deferrable>
constexpr static auto deferrable_part(Deferrable deferrable) {
    if constexpr (decltype(level == none)::value) {
        return empty_sql;
    } else if constexpr (Deferrable::value) {
       return deferrable_sql;
    } else {
        return BOOST_HANA_STRING(" NOT") + deferrable_sql;
    }
}
```

JonasProgrammer replied:
I did it in this way, but this actually fails for the case when level is indeed none, as the comparison operator for `none_t` is called, thus resulting in bool being decltyped.

I fixed this and the other stuff above accordingly, but for the none-case I simply sticked to calling `is_none<T>::value` for `Deferrable`, `Mode` and `Level`.

thed636 replied:
See the comment below :)

thed636 reviewed (changes requested):
It looks like the idea of common `base_async_operation::rebind_initiator` is limiting us here.

Let's try something like this:

```c++
template <typename Operation, typename Initiator>
struct base_async_operation {
    using initiator_type = Initiator;
    using base = base_async_operation;
    initiator_type initiator_;

    constexpr base_async_operation(const Initiator& initiator = Initiator{}) : initiator_(initiator) {}

    constexpr initiator_type get_initiator() const { return initiator_;}

    template <typename InitiatorFactory>
    constexpr auto operator[] (InitiatorFactory&& f) const {
        const Operation& op = static_cast<const Operation&>(*this);
        return op.rebind_initiator(construct_initiator(std::forward<InitiatorFactory>(f), op));
    }
};
```

And each operation should define `rebind_initiator`. In this case, options may become a member of `begin_op`.

```c++
template <typename Initiator, typename Options>
struct begin_op : base_async_operation <begin_op, Initiator> {
    using base = typename begin_op::base;
    Options options_;
    
    begin_op(Initiator initiator, Options options) : base(initiator), options_(options) {}
    //...
    template <typename OtherInitiator>
    constexpr auto rebind_initiator(const OtherInitiator& other) const {
        return begin_op <OtherInitiator, Options>{other, options_};
    }
};
```

JonasProgrammer committed:
Allow customization of transaction initiation statement
- model postgres transaction options in code
- add options parameter to long-form of begin_op::operator()
- add .with_options()-helper to begin_op to control the implicit options
- move rebind_initiator to concrete implementation of basic_async_operations in order to support custom rebinding requirements of the upstream type

Closes #156

thed636 commented on tests/integration/transaction_integration.cpp:
```diff
@@ -54,4 +54,148 @@ TEST(transaction, create_schema_in_transaction_and_rollback_then_table_should_no
 }
 
+TEST(transaction, transaction_level_options_should_work) {
+    // start and commit a transaction with every possible option (mostly to ensure there's no typos in the generated SQL parts)
```
Well, in this case, it is better to rename to

```TEST(transaction, transaction_level_options_should_not_cause_sql_syntax_error)```

And the same things for the others.

thed636 replied:
I think the comments should be removed as you made such a good name for the tests :)

thed636 commented on include/ozo/transaction.h:
```diff
@@ -1,22 +1,38 @@
+    constexpr explicit begin_op(Initiator initiator = {}, Options options = {}) : base(initiator), options_(options) {}
+
+    template <typename T, typename TimeConstraint, typename ConcreteOptions, typename CompletionToken>
+    auto operator() (T&& provider, TimeConstraint t, ConcreteOptions&& options, CompletionToken&& token) const {
```
Well, I think there should be only one method to set-up options to do not provide a misleading user interface. I thought about an additional argument to `operator ()`, but such interface would have a combinatorial explosion of functions in case of trying to keep it consistent. So that's why `with_options()` was the proposal to the interface. Is there any use-case issue or inconvenience to keep both interface solutions?

JonasProgrammer replied:
Well, this is a tough one, but IMHO both interfaces have their use cases. The parameter on `operator ()` is more in line with time constraints and so on.
However, the `with_options` helper is really useful, when you have to use the same isolation level (I need pretty much all of mine to be `SERIALIZABLE`, for example); doing this at connection level is not a good option, when using a shared conn pool. Having to type out the parameter everywhere on the other hand is extra work you can forget at times, when a simple `constexpr auto begin_serial = begin.with_options(...)` makes it clear from the 'function name' alone, that you use your required options, without having to skim through the arg list.

thed636 replied:
Well, I have some arguments against the dedicated version of `operator()`:
1. Yet another version with options has an additional cost (even not so hight for now) for the compilation time.
2. An additional cost to support it.
3. User would ask - why the library does not provide `begin(Provider, Options, Token)` version when I'm using my own timer for the cancelation?

So IMO it is better to rename `with_options` (if it looks like an inconvenient name) to something more laconic than having an additional version of `operator()`.

thed636 commented on include/ozo/transaction.h:
```diff
@@ -75,4 +101,52 @@ struct rollback_op {
+ * @ingroup group-connection-functions
+ * @brief Retrieve's a transactions isolation level
+ *
+ * Retrieve's a transaction's isolation level (that is, an instance of one of the child types of ozo::isolation_level) or an instance of ozo::none_t, if no
```
There is no need to repeat `@brief` section. Also, you can use `scripts/build.sh docs` and look at the result in your browser by opening `build/docs/html/index.html`.

JonasProgrammer replied:
My bad, I was under the impression that @brief is used in the function list and the 'body' in the detailed method description (and as you correctly guessed, I never looked at it in its generated form).

thed636 commented on include/ozo/transaction.h:
```diff
@@ -75,4 +101,52 @@ struct rollback_op {
+ * @return isolation level or none
+ */
+template<typename T>
+constexpr auto get_transaction_isolation_level(T&& transaction) {
```
Since you are using a template parameter `T` instead of `impl::transaction<Ts...>` it is better to make a static assertion which emulates concept assertion about `T`, to provide a better error message. Or just use `const impl::transaction<Ts...>&` as an argument type.

thed636 commented on include/ozo/transaction_options.h:
```diff
@@ -0,0 +1,54 @@
+ * @brief 'type enum' for transaction isolation levels supported by PostgreSQL
+ */
+struct isolation_level {
+    constexpr static hana::type<class serializable_tag> serializable{};
```
It is better to make a short comment about each element like

```c++
constexpr static hana::type<class serializable_tag> serializable{};` //!< Serializable isolation level
```

In other cases no item would be displayed in the documentation. Try to generate it and read it.

thed636 commented on include/ozo/detail/begin_statement_builder.h:
```diff
@@ -0,0 +1,71 @@
+
+    template <typename Level>
+    constexpr static auto level_part(Level level) {
+        auto prefix = BOOST_HANA_STRING(" ISOLATION LEVEL ");
```
[`[[maybe_unused]]`](https://en.cppreference.com/w/cpp/language/attributes/maybe_unused) attribute is better than `void` cast.

`auto` -> `const auto` to explicitly require no modification on this variable.

JonasProgrammer replied:
Yup, sorry, this was a quick&dirty fix because gcc and clang behave differently for -Wunused-variable when it comes to `if constexpr` paths.

thed636 commented on include/ozo/detail/begin_statement_builder.h:
```diff
@@ -0,0 +1,71 @@
+    constexpr static auto deferrable_part(Deferrable) {
+        if constexpr (is_none<std::decay_t<Deferrable>>::value) {
+            return empty_sql;
+        } else if constexpr (auto deferrable_sql = BOOST_HANA_STRING(" DEFERRABLE"); Deferrable::value) {
```
`auto` -> `const auto`

It's up to you, but IMO there are no needs to use is-statement `deferrable_sql` since it is a zero-byte value and fully compile-time thing.

JonasProgrammer replied:
For some reason I misspell 'deferrable' 9/10 times on the first try, so this was more or less added out of spite. Removed for better readability.

thed636 commented on include/ozo/detail/begin_statement_builder.h:
```diff
@@ -0,0 +1,71 @@
+                               begin_statement_builder::level_part(level) +
+                               begin_statement_builder::mode_part(mode) +
+                               begin_statement_builder::deferrable_part(deferrability));
+    }
```
You can try some magic of boost::hana, which allows you to make more generic and linear code:

```c++
using opt = transaction_options;
constexpr auto supported_options = hana::make_tuple(opt::isolation_level, opt::mode, opt::deferrability);
constexpr auto query_prefix = BOOST_HANA_STRING("BEGIN");
//...
template <typename Options>
constexpr static auto build(Options const& options) {
    const auto real_options = hana::filter(supported_options, hana::partial(hana::contains, options));
    const auto strings = hana::transform(real_options, [&](auto& v) { return to_string(v, options[v]);});
    return hana::fold(strings, query_prefix, hana::plus);
}
```

In this case, you do not need to handle `ozo::none`, and can use `to_string` methods instead of `*_part` ones and dispatch it, for example, by the first argument.

thed636 commented on include/ozo/detail/begin_statement_builder.h:
```diff
@@ -0,0 +1,71 @@
+        } else if constexpr (decltype(level == isolation_level::read_uncommitted)::value) {
+            return prefix + BOOST_HANA_STRING("READ UNCOMMITTED");
+        } else {
+            static_assert(std::is_void_v<Level>, "transaction level");
```
It is better to provide a more user-friendly message, like "unexpected transaction level", or something like this, as you did it below. :)

thed636 reviewed (changes requested):
Hi, Jonas!

Thanks for the help! Hope you enjoy the project. I made some comments, and I think we have big progress here. Hana is a really good library for modern C++ metaprogramming which allows to do a lot of things in compile-time in a convenient manner.

thed636 commented on include/ozo/asio.h:
```diff
@@ -211,14 +211,10 @@ struct base_async_operation {
-    }
-
-    template <typename InitiatorFactory>
+    template <typename InitiatorFactory, typename... OperationArgs>
```
It seems like `typename... OperationArgs` is redundant.

thed636 commented on include/ozo/transaction.h:
```diff
@@ -1,22 +1,38 @@
 
-template <typename Initiator>
-struct begin_op : base_async_operation <begin_op, Initiator> {
+template <typename Initiator, typename StatementBuilder, typename Options = decltype(make_options())>
```
Does it really need to point `StatementBuilder` as a template parameter? Does it really need to be customized? I mean what is the customization point it provides?

JonasProgrammer replied:
The original intention was to have it customizable so you could, in theory, roll your own statement builder, e.g. when a newer PostgreSQL version, extension or fork would add additional options for the begin statement.
That being said, this is indeed very unlikely, so I did remove it. If anyone ever needs it, it's easy enough for them to put it back in.

thed636 commented on include/ozo/transaction.h:
```diff
@@ -75,4 +101,50 @@ struct rollback_op {
+ * `ozo::none` if no level was explicitly set.
+ *
+ * @param transaction transaction to get the isolation level from
+ * @return isolation level or none
```
It is better to write something like this

```
@return ozo::isolation_level --- if the isolation level has been set
@return ozo::none --- if no isolation level has been specified for the transaction
```

In this case, necessary links should be generated by Doxygen.

thed636 commented on include/ozo/transaction_options.h:
```diff
@@ -0,0 +1,61 @@
+
+/**
+ * @ingroup group-connection-types
+ * @brief 'type enum' for transaction isolation levels supported by PostgreSQL
```
I'd like to ask you to provide for this and the other options a link for the related PostgreSQL documentation here, to allow a user easily navigate to it.

thed636 commented on tests/detail/begin_statement_builder.cpp:
```diff
@@ -0,0 +1,211 @@
+
+struct my_true_t { constexpr static bool value = true; } my_true;
+
+TEST(begin_statement_builder, should_allow_integral_constants_for_deferrability) {
```
I'm not sure about this test is necessary. You provide `deferrable` expectations in the previous test. I do not think that check for support custom true or false type is really needed.

thed636 commented on include/ozo/detail/begin_statement_builder.h:
```diff
@@ -0,0 +1,69 @@
+            return prefix + BOOST_HANA_STRING("READ COMMITTED");
+        } else if constexpr (decltype(level == isolation_level::read_uncommitted)::value) {
+            return prefix + BOOST_HANA_STRING("READ UNCOMMITTED");
+        } else {
```
It is not really mandatory, but since `isolation_level` members are `hana::type` you can try to use [hana::map](https://www.boost.org/doc/libs/1_67_0/libs/hana/doc/html/structboost_1_1hana_1_1map.html) here for dispatching. I think you could like the results. :)

thed636 reviewed (changes requested):
Hi, Jonas!

Sorry for the delay caused by C++Con attend. I made some comments on this version of PR. It much better now, but we still have to solve some problems. Thanks for your help and patient. :)

With best regards, 
Sergei.
